### PR TITLE
Shopify: manage connected/disconnected states

### PR DESCRIPTION
**Pull request by cbernat**
Shopify section:
- Add Shopify client and shopify hardcoded client
- Add tests
- Add view with title only, access via /#/integrations/shopify
- Add header without section

image

https://cdn.fromdoppler.com/doppler-webapp/build883/#/integrations/shopify

### CODE REVIEW

File: src/services/shopify-client.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+export interface ShopifyConnection {
+  shop_name: string;
+  synchronization_date: Date | null;
+  list: SubscriberList;
+  connected: boolean;
+}
+
+export interface ShopifyClient {
+  getShopifyData(email: string, apiKey: string): Promise<ShopifyConnection>;
+  disconnect(email: string, apikey: string): Promise<boolean>;
```
Comment (andresmoschini) [inline src/services/shopify-client.ts]: Why does it return a boolean? If the idea is return _false_ when there is an error, I would suggest ussing something like [EmptyResult<TError>](https://github.com/FromDoppler/doppler-webapp/blob/master/src/services/doppler-legacy-client.ts#L19).

Maybe those classes could be moved to a common place.

File: src/services/shopify-client.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+export interface ShopifyConnection {
+  shop_name: string;
+  synchronization_date: Date | null;
+  list: SubscriberList;
+  connected: boolean;
+}
+
+export interface ShopifyClient {
+  getShopifyData(email: string, apiKey: string): Promise<ShopifyConnection>;
```
Comment (andresmoschini) [inline src/services/shopify-client.ts]: If I am not wrong, there could be more than a shop connected, what do you think about:

```suggestion
  getShopifyData(email: string, apiKey: string): Promise<ShopifyConnection[]>;
```

File: src/services/shopify-client.doubles.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
+import {
```
Comment (andresmoschini) [inline src/services/shopify-client.doubles.ts]: Si no me equivoco, ni _ShopifyClient_, ni _HardcodedShopifyClient_ se están usando, tal vez convenga mergear solo la ruta y la página.

File: src/components/Integrations/Shopify/Shopify.js [hunk 4]
```diff
@@ -0,0 +1,13 @@
+import React from 'react';
+import { Helmet } from 'react-helmet';
+
+const Shopify = () => (
+  <>
+    <Helmet>
+      <title>Doppler | Shopify </title>
+    </Helmet>
+    <div> Control panel page for shopify</div>
```
Comment (andresmoschini) [inline src/components/Integrations/Shopify/Shopify.js]:
```suggestion
    <div>Control panel page for Shopify</div>
```

File: src/components/Integrations/Shopify/Shopify.js [hunk 5]
```diff
@@ -0,0 +1,13 @@
+import React from 'react';
+import { Helmet } from 'react-helmet';
+
+const Shopify = () => (
+  <>
+    <Helmet>
+      <title>Doppler | Shopify </title>
```
Comment (andresmoschini) [inline src/components/Integrations/Shopify/Shopify.js]:
```suggestion
      <title>Doppler | Shopify</title>
```

File: src/services/shopify-client.ts [hunk 6]
```diff
@@ -0,0 +1,25 @@
+export enum SubscriberListState {
+  ready,
+  synchronizing_contacts,
```
Comment (andresmoschini) [inline src/services/shopify-client.ts]: What do you think about:

```suggestion
  synchronizingContacts,
```

File: src/services/shopify-client.ts [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+}
+
+export interface SubscriberList {
+  name: string;
+  id: number;
+  amount_subscribers: number;
```
Comment (andresmoschini) [inline src/services/shopify-client.ts]: What do you think about:

```suggestion
  subscribersCount: number;
```

File: src/services/shopify-client.ts [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+  state: SubscriberListState;
+}
+
+export interface ShopifyConnection {
+  shop_name: string;
```
Comment (andresmoschini) [inline src/services/shopify-client.ts]: What do you think about:

```suggestion
  shopName: string;
```

File: src/services/shopify-client.ts (hunk 1 above)
Comment (cbernat) [reply src/services/shopify-client.ts]: About the EmptyResult, yes maybe it's better. Which classes? I don't get you with this.

File: src/services/shopify-client.doubles.ts (hunk 3 above)
Comment (cbernat) [reply src/services/shopify-client.doubles.ts]: Puede ser, o lo que me comentaste por chat, podría subir solamente que muestre si esta integrado o no. Usando el client.

File: src/services/shopify-client.ts (hunk 8 above)
Comment (cbernat) [reply src/services/shopify-client.ts]: We don't use underscores anywhere right?

Comment (andresmoschini) [reply src/services/shopify-client.ts]: Mmm, no, in general, we are using camel case.

**Comment by cbernat** [comment]
Test this please.

File: src/components/Integrations/Shopify/Shopify.js [hunk 9]
```diff
@@ -1,16 +1,33 @@
...
-  </>
-);
+const shopifyClient = new HardcodedShopifyClient();
+
+const Shopify = () => {
+  const [shopName, setShopName] = useState('');
+  useEffect(() => {
+    const getData = async () => {
+      const result = await shopifyClient.getShopifyData();
+      if (result.length) {
```
Comment (andresmoschini) [inline src/components/Integrations/Shopify/Shopify.js]: What happens in the other case?

File: src/components/Integrations/Shopify/Shopify.js [hunk 10]
```diff
@@ -1,16 +1,33 @@
...
+        setShopName(result[0].shopName);
+      }
+    };
+    getData();
+  }, []);
```
Comment (andresmoschini) [inline src/components/Integrations/Shopify/Shopify.js]: Why do you need `getData` function? I think that it will work in the same way:

```js
  useEffect(() => {
      const result = await shopifyClient.getShopifyData();
      if (result.length) {
        setShopName(result[0].shopName);
      }
  }, []);
```

File: src/components/Integrations/Shopify/Shopify.js [hunk 11]
```diff
@@ -1,16 +1,33 @@
...
+
+  return (
+    <>
+      <Helmet>
+        <title>Doppler | Shopify</title>
+      </Helmet>
+      <section style={{ width: '100%', padding: '60px 30px' }}>
+        Panel de control | Integraciones y preferencias avanzadas
+        <h1>Panel de Control - Shopify</h1>
+        {shopName.length ? 'Shopify conectado a tienda: ' + shopName : 'Shopify desconectado'}
```
Comment (andresmoschini) [inline src/components/Integrations/Shopify/Shopify.js]: I think that you are missing some possible states:
- [ ] Undetermined (when ajax call did not finished yet)
- [x] When there is not a shop
- [x] When there is a shop
- [ ] When there was an error in ajax call

File: src/services/shopify-client.doubles.ts [hunk 12]
```diff
@@ -15,12 +10,12 @@ const fakeData = [
...
       state: SubscriberListState.ready,
-    }
-  }
+    },
+  },
 ];
 
 export class HardcodedShopifyClient implements ShopifyClient {
-  public async getShopifyData(email: string, apiKey: string): Promise<ConnectedShops[]> {
+  public async getShopifyData(): Promise<ConnectedShops[]> {
```
Comment (andresmoschini) [inline src/services/shopify-client.doubles.ts]: In my opinion, the result type should be a wrapper that takes into account errors.

Comment (cbernat) [reply src/services/shopify-client.doubles.ts]: Yes you are right, I will make a small refactor to reuse dopplerLegacyClient results. In a separate commit.

File: src/components/Integrations/Shopify/Shopify.js (hunk 9 above)
Comment (cbernat) [reply src/components/Integrations/Shopify/Shopify.js]: You are clearly not connected, without shops that is.

File: src/components/Integrations/Shopify/Shopify.js (hunk 10 above)
Comment (cbernat) [reply src/components/Integrations/Shopify/Shopify.js]: I need to tell the function is async somewhere 
image

Comment (cbernat) [reply src/components/Integrations/Shopify/Shopify.js]: I've tried adding async into useEffect hook but it also gives error, in this post https://www.robinwieruch.de/react-hooks-fetch-data/#react-hooks-programmatically says that the ways is with a function.

Comment (andresmoschini) [reply src/components/Integrations/Shopify/Shopify.js]: Right, I would like to understand why the `useEffect(async () => {` does not work, do you have a screenshot of the error? and if it does not work, could you explain to me in a short way? (I have still pending reading the article)

File: src/components/Integrations/Shopify/Shopify.js (hunk 9 above)
Comment (andresmoschini) [reply src/components/Integrations/Shopify/Shopify.js]: In that case, I think that you should make it explicit.

Maybe something like this:

```js
  const [shopName, setShopName] = useState(null);
  // shopName === null would mean "loading"
  useEffect(() => {
    const getData = async () => {
      const result = await shopifyClient.getShopifyData();
      if (result.length) {
        setShopName(result[0].shopName);
        // shopName === something would mean "there is a shop"
      } else {
        setShopName('');
        // shopName === '' would mean "there is not a shop"
    };
    getData();
  }, []);
```

I think that it could work, but maybe it would be better having more than a _state_ in order to make it more explicit. Or having something different, for example an object:


```js
  cont _thereIsNoShop = {};
  // . . .
  const [shop, setShop] = useState(null);
  // shop === null would mean "loading"
  useEffect(() => {
    const getData = async () => {
      const result = await shopifyClient.getShopifyData();
      if (result.length) {
        setShopName(result[0]);
        // shop would contains "the shop".
      } else {
        setShop(_thereIsNoShop);
        // shop === _thereIsNoShop would mean that there is no job
    };
    getData();
  }, []);
```

File: src/components/Integrations/Shopify/Shopify.js (hunk 10 above)
Comment (cbernat) [reply src/components/Integrations/Shopify/Shopify.js]: According to the documentation every function annotated with async returns an implicit promise: “The async function declaration defines an asynchronous function, which returns an AsyncFunction object. An asynchronous function is a function which operates asynchronously via the event loop, using an implicit Promise to return its result. “. However, an effect hook should return *nothing or a clean up function*. That’s why you may see the following warning in your developer console log: 07:41:22.910 index.js:1452 Warning: useEffect function must return a cleanup function or nothing. Promises and useEffect(async () => …) are not supported, but you can call an async function inside an effect.. That’s why using async directly in the useEffect function isn’t allowed.

Comment (andresmoschini) [reply src/components/Integrations/Shopify/Shopify.js]: Oh, I understand now, thank you very much for the clear explanation!

File: src/components/Integrations/Shopify/Shopify.js [hunk 13]
```diff
@@ -7,13 +7,25 @@ const shopifyClient = new HardcodedShopifyClient();
...
+          setShopName('');
+        } else {
+          if (result.expectedError && result.expectedError.cannotConnectToAPI) {
+            setError(
+              'Error: No hemos podido conectar con la Api de Shopify, vuelve a intentar luego.',
+            );
+          } else {
+            setError('Error: Error inesperado.');
+          }
+        }
```
Comment (andresmoschini) [inline src/components/Integrations/Shopify/Shopify.js]: que opinás de reescribirlo de esta manera para reducir la anidación?

```js
        } else  if (result.expectedError && result.expectedError.cannotConnectToAPI) {
          setError(
            'Error: No hemos podido conectar con la Api de Shopify, vuelve a intentar luego.',
          );
        } else {
          setError('Error: Error inesperado.');
        }
      }
```

File: src/services/shopify-client.ts [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+
+export interface ConnectedShops {
+  shopName: string;
+  synchronization_date: Date | null;
+  list: SubscriberList;
+}
+
+export type ShopifyErrorResult = { cannotConnectToAPI: true };
+export interface ShopifyClient {
+  getShopifyData(): Promise<ConnectedShops[] | EmptyResult<ShopifyErrorResult>>;
```
Comment (andresmoschini) [inline src/services/shopify-client.ts]: Creo que el tipo de resultado debería ser `Result<ConnectedShops[], ShopifyErrorResult>`

```suggestion
  getShopifyData(): Promise<Result<ConnectedShops[], ShopifyErrorResult>>;
```

Comment (cbernat) [reply src/services/shopify-client.ts]: This should behave in the same way right?

Comment (andresmoschini) [reply src/services/shopify-client.ts]: So so, it will enforce you to wrap the result and include the `success: true` field, like here:

https://github.com/FromDoppler/doppler-webapp/blob/7b5399b8ec18d8fbbf94032ab75e2f3bc7428af9/src/services/doppler-legacy-client.ts#L321-L328

File: src/components/Integrations/Shopify/Shopify.js (hunk 13 above)
Comment (andresmoschini) [reply src/components/Integrations/Shopify/Shopify.js]: Ah, hay otra anidación más. En general, este tipo de lógica se pueden leer mucho mejor cuando hay un solo nivel de anidación.

En este caso yo sugeriría la siguiente estructura:

```js
      const result = await shopifyClient.getShopifyData();
      if (result.success) { 
        // ahora mismo no funciona porque el tipo no es Result
        // . . .
      } else if (!result.value.length) { 
        // el campo value tambiés es parte del tipo Result
        // Tal vez incluso se pueda definir un tipo diferente de resultado para
        // cuando no hay items
        // . . .
      } else if (result.expectedError && result.expectedError.cannotConnectToAPI) {
        // . . .
      } else {
        // . . .
      }
```

Comment (cbernat) [reply src/components/Integrations/Shopify/Shopify.js]: I get your point.

File: src/services/shopify-client.ts (hunk 14 above)
Comment (cbernat) [reply src/services/shopify-client.ts]: I've tried but for some reason it gives error. 
image

Comment (andresmoschini) [reply src/services/shopify-client.ts]: It is not empty any more hehe

Comment (andresmoschini) [reply src/services/shopify-client.ts]: please, let me know if it is not clear, we can talk about it.

Comment (cbernat) [reply src/services/shopify-client.ts]: Yeah I've noticed it. I'm using `Result` now locally.

File: src/components/Integrations/Shopify/Shopify.js [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+      <Helmet title={'Doppler | Shopify'} />
+      <section style={{ width: '100%', padding: '60px 30px' }}>
+        Panel de control | Integraciones y preferencias avanzadas
+        <h1>Panel de Control - Shopify</h1>
+        {isLoading ? (
+          <Loading />
+        ) : error ? (
+          error
+        ) : isConnected ? (
+          shops.map((shop) => 'Shopify conectado a tienda: ' + shop.shopName)
```
Comment (andresmoschini) [inline src/components/Integrations/Shopify/Shopify.js]: If I am not wrong, it returns an array, what do you think about one of these options?

```suggestion
          `Shopify conectado a tienda: ${shops[0].shopName}`
```

Or:

```suggestion
          <p>Conectado a:</p>
          <ul>
            {shops.map((shop) => (<li>{shop.shopName}</li>)}
          </ul>
```

Comment (cbernat) [reply src/components/Integrations/Shopify/Shopify.js]: I like the second option best.

File: src/components/Header/Header.js [hunk 16]
```diff
@@ -4,14 +4,15 @@ import HeaderMessages from './HeaderMessages/HeaderMessages';
 import HeaderUserMenu from './HeaderUserMenu/HeaderUserMenu';
 import { FormattedMessage } from 'react-intl';
 
-const Header = ({ userData: { user, nav, alert } }) => {
+const Header = ({ userData: { user, nav, alert }, location }) => {
+  const inactiveSection = location.pathname.match(/^\/integrations\/*/) !== null;
```
Comment (andresmoschini) [inline src/components/Header/Header.js]: Maybe using deconstruction, the dependency will be more explicit:

```js
const Header = ({ userData: { user, nav, alert }, location: { pathname }}) => {
  const inactiveSection = pathname.match(/^\/integrations\/*/) !== null; 
```

File: src/components/form-helpers/captcha-utils.tsx [hunk 17]
```diff
@@ -6,7 +6,7 @@ type CaptchaUtils = Readonly<
   [
     (props: any) => JSX.Element,
     () => Promise<CaptchaVerificationResult>,
-    React.RefObject<ReCAPTCHA>,
+    React.RefObject<ReCAPTCHA>
```
Comment (andresmoschini) [inline src/components/form-helpers/captcha-utils.tsx]: Please revert this change, it seems that it is being generated by a wrong _Prettier_ version.

Try this:

```bash
$ rm -rf  node_modules
$ yarn
$ yarn prettier
```

File: src/services/shopify-client.doubles.ts [hunk 18]
```diff
@@ -20,37 +20,10 @@ const oneShop = [
...
-
 export class HardcodedShopifyClient implements ShopifyClient {
   public async getShopifyData(): Promise<Result<ConnectedShop[], ShopifyErrorResult>> {
     console.log('getShopifyData');
     await timeout(1500);
-    //return {success: false, message: 'Some random error'};
-    //return {success: false, expectedError: {cannotConnectToAPI: true}};
-    //return { success: true, value: twoShops };
     return { success: true, value: oneShop };
-    //return { success: true, value:[]};
```
Comment (andresmoschini) [inline src/services/shopify-client.doubles.ts]: What do you think about keeping these commented lines? I think that they could help us to test manually.

**Comment by andresmoschini** [comment]
Muy bueno! creo que ya se le podría ir quitando el WIP y mergeando, ¿que opinás?

**Comment by cbernat** [comment]
> Muy bueno! creo que ya se le podría ir quitando el WIP y mergeando, ¿que opinás?

Tiene como veinte millones de commits, lo emprolijo en otro PR mejor :)

**Comment by andresmoschini** [comment]
Oh, remember to also take into account Travis errors (We are still not paying attention to them in our Jenkins task):

image

image

**Comment by cbernat** [comment]
Test this please

File: src/doppler-types.ts [hunk 19]
```diff
@@ -0,0 +1,10 @@
+// TODO: move it a common place if it will be reused
```
Comment (andresmoschini) [inline src/doppler-types.ts]: Esta línea creo que ya no es necesaria.

**Comment by andresmoschini** [comment]
:shipit:

**Comment by GusBaamonde** [comment]
This is weird, this later I check it

shopy

File: src/components/Header/Header.js [hunk 20]
```diff
@@ -4,14 +4,19 @@ import HeaderMessages from './HeaderMessages/HeaderMessages';
...
     <div>
       {alert ? <HeaderMessages alert={alert} user={user} /> : null}
       {/* //TODO: Refactor backend to send proper active values. Class 'header-is-active' must be removed */}
       <header
         className={
-          user.clientManager ? 'header-main header-open dp-header--cm' : 'header-main header-open'
+          inactiveSection
+            ? 'header-main'
+            : user.clientManager
+            ? 'header-main header-open dp-header--cm'
```
Comment (fcoronelmakingsense) [inline src/components/Header/Header.js]: cant do this in a better way? because repeat 'header main' always

File: src/components/Integrations/Shopify/Shopify.js [hunk 21]
```diff
@@ -0,0 +1,59 @@
...
+      if (result.success && result.value.length) {
+        setIsConnected(true);
+        setShops(result.value);
+      } else if (
+        !result.success &&
+        result.expectedError &&
+        result.expectedError.cannotConnectToAPI
+      ) {
+        setError('Error: No hemos podido conectar con la Api de Shopify, vuelve a intentar luego.');
+      } else if (!result.success) {
```
Comment (fcoronelmakingsense) [inline src/components/Integrations/Shopify/Shopify.js]: maybe its better ask first this and then dont need to repeat ask for `!result.success`

File: src/components/Integrations/Shopify/Shopify.js [hunk 22]
```diff
@@ -0,0 +1,59 @@
...
+      setIsLoading(false);
+    };
+    getData();
+  }, [shopifyClient]);
+
+  return (
+    <>
+      <Helmet title={'Doppler | Shopify'} />
+      <section style={{ width: '100%', padding: '60px 30px' }}>
+        Panel de control | Integraciones y preferencias avanzadas
```
Comment (fcoronelmakingsense) [inline src/components/Integrations/Shopify/Shopify.js]: dont use i18n texts?

File: src/components/Integrations/Shopify/Shopify.js [hunk 23]
```diff
@@ -0,0 +1,59 @@
...
+      }
+      setIsLoading(false);
+    };
+    getData();
+  }, [shopifyClient]);
+
+  return (
+    <>
+      <Helmet title={'Doppler | Shopify'} />
+      <section style={{ width: '100%', padding: '60px 30px' }}>
```
Comment (fcoronelmakingsense) [inline src/components/Integrations/Shopify/Shopify.js]: its better use css class, what do you think?

Comment (GusBaamonde) [reply src/components/Integrations/Shopify/Shopify.js]: I think it's great, but it's something to fix in another PR. Then we see it.  ;)

File: src/services/shopify-client.doubles.ts [hunk 24]
```diff
@@ -0,0 +1,56 @@
...
+    list: {
+      name: 'MyList',
+      id: 1251,
+      amountSubscribers: 2,
+      state: SubscriberListState.ready,
+    },
+  },
+];
+
+/*const twoShops = [
```
Comment (fcoronelmakingsense) [inline src/services/shopify-client.doubles.ts]: need this code?

File: src/services/shopify-client.doubles.ts [hunk 25]
```diff
@@ -0,0 +1,56 @@
...
+      state: SubscriberListState.ready,
+    },
+  },
+];*/
+
+export class HardcodedShopifyClient implements ShopifyClient {
+  public async getShopifyData(): Promise<Result<ConnectedShop[], ShopifyErrorResult>> {
+    console.log('getShopifyData');
+    await timeout(1500);
+    //return {success: false, message: 'Some random error'};
```
Comment (fcoronelmakingsense) [inline src/services/shopify-client.doubles.ts]: need all the comments? what do you think @andresmoschini

File: src/components/Integrations/Shopify/Shopify.js (hunk 22 above)
Comment (cbernat) [reply src/components/Integrations/Shopify/Shopify.js]: Not for now. This was meant to be very simple.

File: src/components/Integrations/Shopify/Shopify.js (hunk 23 above)
Comment (cbernat) [reply src/components/Integrations/Shopify/Shopify.js]: This is temporal, next I will be adding styles already merged into ui library, check https://cdn.fromdoppler.com/doppler-ui-library/latest/integrations.html

File: src/services/shopify-client.doubles.ts (hunk 24 above)
Comment (cbernat) [reply src/services/shopify-client.doubles.ts]: Yes, for quick testing purposes.

File: src/doppler-types.ts (hunk 19 above)
Comment (cbernat) [reply src/doppler-types.ts]: > This is weird, this later I check it
> 
> shopy



> This is weird, this later I check it
> 
> shopy

I don't see how to reproduce this, can you tell me please?

**Comment by cbernat** [comment]
> This is weird, this later I check it
> 
> shopy

I don't get how you reproduce this, would you mind telling me how?

File: src/services/shopify-client.doubles.ts (hunk 25 above)
Comment (cbernat) [reply src/services/shopify-client.doubles.ts]: I've added this again, at @andresmoschini request. For quick testing purposes.

**Comment by GusBaamonde** [comment]
> No entiendo cómo reproduces esto, ¿te importaría decirme cómo?
Yes! You enter https://cdn.fromdoppler.com/doppler-webapp/build883/#/integrations/shopify
and unfold the user menu

File: src/components/Integrations/Shopify/Shopify.js (hunk 21 above)
Comment (andresmoschini) [reply src/components/Integrations/Shopify/Shopify.js]: The alternative are _nested ifs_, and they tend to be less clear. Intendation is our enemy! :P

File: src/services/shopify-client.doubles.ts (hunk 25 above)
Comment (andresmoschini) [reply src/services/shopify-client.doubles.ts]: These comments are really useful for testing weird scenarios in the local environment.

File: src/components/Integrations/Shopify/Shopify.js (hunk 23 above)
Comment (fcoronelmakingsense) [reply src/components/Integrations/Shopify/Shopify.js]: can add html comment? to dont forget change

File: src/services/shopify-client.doubles.ts (hunk 25 above)
Comment (fcoronelmakingsense) [reply src/services/shopify-client.doubles.ts]: thanks for answer!

File: src/components/Integrations/Shopify/Shopify.js (hunk 21 above)
Comment (fcoronelmakingsense) [reply src/components/Integrations/Shopify/Shopify.js]: maybe can use something like that too.
```js
if(result.expectedError &&
        result.expectedError.cannotConnectToAPI){}
    else if(!result.sucess){}
    else if(result.value.length){}
```js

Comment (cbernat) [reply src/components/Integrations/Shopify/Shopify.js]: I will try this out. Thanks!

Comment (cbernat) [reply src/components/Integrations/Shopify/Shopify.js]: I went out perfect! Thanks!

File: src/components/Integrations/Shopify/Shopify.js (hunk 23 above)
Comment (cbernat) [reply src/components/Integrations/Shopify/Shopify.js]: @fcoronelmakingsense I can't seem to add html commented code in between. But I will remember, my next pr will be about applying styles in the library.

File: src/components/Integrations/Shopify/Shopify.js [hunk 26]
```diff
@@ -0,0 +1,55 @@
...
+    const getData = async () => {
+      const result = await shopifyClient.getShopifyData();
+      if (result.expectedError && result.expectedError.cannotConnectToAPI) {
+        setError('Error: No hemos podido conectar con la Api de Shopify, vuelve a intentar luego.');
+      } else if (!result.success) {
+        setError('Error: Error inesperado.');
+      } else if (result.value.length) {
+        setIsConnected(true);
+        setShops(result.value);
+      }
```
Comment (andresmoschini) [inline src/components/Integrations/Shopify/Shopify.js]: Creo que falta esto:

```suggestion
      } else {
        setIsConnected(false);
        setShops(result.value);
      }
```

Comment (cbernat) [reply src/components/Integrations/Shopify/Shopify.js]: That's not quite right, if I remove the condition from this last if in the case that is disconnected, it fails one test. 
```
return { success: true, value:[]};
```
If there is no correct data change through out the if, by default is disconnected.

Comment (cbernat) [reply src/components/Integrations/Shopify/Shopify.js]: I don't need an extra else, because is disconnected by default. Perhaps it's confusing...

Comment (andresmoschini) [reply src/components/Integrations/Shopify/Shopify.js]: In general, having deterministic functions that do not depend on external context (original value) is a good thing.

Take into account that this function will be executed whenever React wants. So, maybe this function could be executed twice. 

Suppose this scenario: the current user is changed, I do not know, a login in another page, or something happens with current shopify integration, before there was a shop, and now there is not a shop, before was connected, and now it is not.

You are missing this scenario here.
